contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,978 @@
+        nonReentrant
+    {
+        // to add liquidity to the clPool
+        IERC20(WETH).safeApprove(address(positionManager), type(uint256).max);
+        IERC20(OETHb).safeApprove(address(positionManager), type(uint256).max);
+        // to be able to rebalance using the swapRouter
+        IERC20(WETH).safeApprove(address(swapRouter), type(uint256).max);
+        IERC20(OETHb).safeApprove(address(swapRouter), type(uint256).max);
+    }
+
+    /***************************************
+            Balances and Fees
+    ****************************************/
+
+    /**
+     * @dev Get the total asset value held in the platform
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
+        external
+        view
+        override
+        returns (uint256)
+    {
+        require(_asset == WETH, "Only WETH supported");
+
+        // we could in theory deposit to the strategy and forget to call rebalance in the same
+        // governance transaction batch. In that case the WETH that is on the strategy contract
+        // also needs to be accounted for.
+        uint256 _wethBalance = IERC20(WETH).balanceOf(address(this));
+        // just paranoia check, in case there is OETHb in the strategy that for some reason hasn't
+        // been burned yet.
+        uint256 _oethbBalance = IERC20(OETHb).balanceOf(address(this));
+        return underlyingAssets + _wethBalance + _oethbBalance;
+    }
+
+    /**
+     * @dev Returns the balance of both tokens in a given position (excluding fees)
+     * @return _amountWeth Amount of WETH in position
+     * @return _amountOethb Amount of OETHb in position
+     */
+    function getPositionPrincipal()
+        public
+        view
+        returns (uint256 _amountWeth, uint256 _amountOethb)
+    {
+        if (tokenId == 0) {
+            return (0, 0);
+        }
+
+        uint160 _sqrtRatioX96 = getPoolX96Price();
+        (_amountWeth, _amountOethb) = helper.principal(
+            positionManager,
+            tokenId,
+            _sqrtRatioX96
+        );
+    }
+
+    /**
+     * @notice Returns the current pool price in X96 format
+     * @return _sqrtRatioX96 Pool price
+     */
+    function getPoolX96Price() public view returns (uint160 _sqrtRatioX96) {
+        (_sqrtRatioX96, , , , , ) = clPool.slot0();
+    }
+
+    /**
+     * @notice Returns the current active trading tick of the underlying pool
+     * @return _currentTick Current pool trading tick
+     */
+    function getCurrentTradingTick() public view returns (int24 _currentTick) {
+        (, _currentTick, , , , ) = clPool.slot0();
+    }
+
+    /**
+     * @notice Returns the amount of liquidity in the contract's LP position
+     * @return _liquidity Amount of liquidity in the position
+     */
+    function _getLiquidity() internal view returns (uint128 _liquidity) {
+        if (tokenId == 0) {
```

Comment:
Agreed that it's not reachable. I think it's still okay for an internal method to check its own prerequisites. If someone else had a token 0 with some liquidity on it, we could be surprised one day when making code changes.

Reply:
100% agree.

Reply:
Good point, though this is sort of a paranoia check. Just in case we make a change in the future to use the `_getLiquidity` by some other function, it is nice to have a check contained within the function as well. Otherwise we would need to add a comment to the function notifying anyone calling it that the `toenId != 0` check needs to be applied outside the function as well. 
Seems easier just to check for it again. 

It is on the other hand not very gas efficient. But we can worry about that once we deploy it to a chain that has high gas cost

Reply:
Ah sorry saying twice what has been already said. I didn't refresh the page and haven't Daniel's response yet.